In [12]:
#Task 1.1
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

athlete_selection = pd.read_csv('/kaggle/input/lab3ccai/AthleteSelection.csv')

athlete_selection.set_index("Athlete", inplace=True)

# Check if the index is having unique values
print("\nIs the Pandas index having unique values?\n",athlete_selection.index.is_unique)

athlete_selection.head(5)




Is the Pandas index having unique values?
 True


Speed  Agility  Selected
Athlete                          
x1        2.50     6.00         0
x2        3.75     8.00         0
x3        2.25     5.50         0
x4        3.25     8.25         0
x5        2.75     7.50         0

In [13]:
#Task 1.2
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

athlete_selection = pd.read_csv('/kaggle/input/lab3ccai/AthleteSelection.csv')

# Create a new variable "names" that contains the DataFrame indexes
names = athlete_selection.index

# Print the "names" variable
print(names)

RangeIndex(start=0, stop=20, step=1)


In [14]:
#Task 1.3
import pandas as pd
import numpy as np

athlete_selection = pd.read_csv('/kaggle/input/lab3ccai/AthleteSelection.csv')

X = np.array(athlete_selection.iloc[:, 1:5])
Y = np.array(athlete_selection.iloc[:, -1])

print(X[0,0])

2.5


In [15]:
#Task 1.4
from sklearn.neighbors import NearestNeighbors
import numpy as np

# Load the data
data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteSelection.csv', delimiter=',', skip_header=1)
X = data[:, 1:3] # Features
y = data[:, 3] # Labels

# Create and fit the model
model = NearestNeighbors(n_neighbors=2, radius=0.4)
model.fit(X)

NearestNeighbors(n_neighbors=2, radius=0.4)

In [16]:
#Task 1.5
# Get the parameters of the model
params = model.get_params()
print(params)

{'algorithm': 'auto', 'leaf_size': 30, 'metric': 'minkowski', 'metric_params': None, 'n_jobs': None, 'n_neighbors': 2, 'p': 2, 'radius': 0.4}


In [17]:
#Task 1.6
from sklearn.neighbors import NearestNeighbors
import numpy as np

# Load the data
data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteSelection.csv', delimiter=',', skip_header=1)
X = data[:, 1:3] # Features
y = data[:, 3] # Labels

# Create and fit the model
model = NearestNeighbors(n_neighbors=2, radius=0.4)
model.fit(X)

# Find the k nearest neighbors of q1 and q2
q1 = [3.25,8.25]
q2 = [0.2, 3.3]
distances, indices = model.kneighbors([q1, q2])

# Stack distances and indices vertically
results = np.vstack((distances, indices))

# Print the results as a matrix
print("Distances and Indices of the nearest neighbors:")
print(results)

Distances and Indices of the nearest neighbors:
[[ 0.          0.55901699]
 [ 2.22036033  2.80044639]
 [ 3.          1.        ]
 [10.          7.        ]]


In [18]:
#Task 1.7
from sklearn.neighbors import NearestNeighbors
import numpy as np

# Load the data
data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteSelection.csv', delimiter=',', skip_header=1)
X = data[:, 1:3] # Features
y = data[:, 3] # Labels

# Create and fit the model
model = NearestNeighbors(n_neighbors=2, radius=0.4)
model.fit(X)

# Find the k nearest neighbors of q1 and q2
q1 = [3.25,8.25]
q2 = [0.2, 3.3]
distances, indices = model.kneighbors([q1, q2])

# Print the results
print(f"The 2 nearest neighbors of {q1} are:")
for i, index in enumerate(indices[0]):
    print(f"{i+1}. Point {index} at distance {distances[0][i]}")
print(f"\nThe 2 nearest neighbors of {q2} are:")
for i, index in enumerate(indices[1]):
    print(f"{i+1}. Point {index} at distance {distances[1][i]}")

The 2 nearest neighbors of [3.25, 8.25] are:
1. Point 3 at distance 0.0
2. Point 1 at distance 0.5590169943749475

The 2 nearest neighbors of [0.2, 3.3] are:
1. Point 10 at distance 2.2203603311174516
2. Point 7 at distance 2.800446392988089


q = [5.0,7.5]
q3n = athlete_neigh.kneighbors([q], n_neighbors = 3)[1][0]
for n in q3n:
  print(names[n])


This code finds the 3 nearest neighbors of the point q using a NearestNeighbors model called athlete_neigh. The kneighbors method returns two arrays: the first contains the distances to the nearest neighbors, and the second contains the indices of the nearest neighbors. In this case, only the indices are used, which are accessed using [1][0]. The indices are then used to access the corresponding names from a list called names, and these names are printed out.

In [19]:
#Task 1.8
from sklearn.neighbors import KNeighborsClassifier
import numpy as np

# Load the data
data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteSelection.csv', delimiter=',', skip_header=1)
X = data[:, 1:3] # Features
y = data[:, 3] # Labels

# Create and fit the model
model = KNeighborsClassifier(n_neighbors=3)
model.fit(X, y)

KNeighborsClassifier(n_neighbors=3)

In [20]:
#Task 1.9
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import numpy as np

# Load the data
data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteSelection.csv', delimiter=',', skip_header=1)
X = data[:, 1:3] # Features
y = data[:, 3] # Labels

# Create and fit the model
model = KNeighborsClassifier(n_neighbors=3)
model.fit(X, y)

# Make predictions on the training data
y_pred = model.predict(X)
# Print the predicted labels and the actual labels
print("Predicted labels:", y_pred)
print("Actual labels:", y)

# Calculate the accuracy
accuracy = accuracy_score(y, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Predicted labels: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1.]
Actual labels: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1.]
Accuracy: 1.00


#Task 1.10

Although the model's accuracy may be high, there are multiple variables that can influence its performance. A few potential factors that could explain why a KNeighborsClassifier model is precise include: the data is appropriate for a nearest neighbors methodology, the K value is suitable, and the features used to describe the data are informative and capable of distinguishing. It is challenging to establish precisely why the model may be accurate without additional insights about the data and the problem at hand.

TASK 2:

In [21]:
# Task 2.1
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import accuracy_score
import numpy as np

# 1. Load the test data
test_data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteTest.csv', delimiter=',', skip_header=1)
X_test = test_data[:, 1:3] # Features
y_test = test_data[:, 3] # Labels

# Load the training data
train_data = np.genfromtxt('/kaggle/input/lab3ccai/AthleteTest.csv', delimiter=',', skip_header=1)
X_train = train_data[:, 1:3] # Features
y_train = train_data[:, 3] # Labels

In [22]:
# Task 2.2 Create and fit the KNeighborsClassifier model with K=3
model = KNeighborsClassifier(n_neighbors=3)
model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [23]:
# Task 2.3 Evaluate the model on the test data
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy of KNeighborsClassifier with K=3: {accuracy:.2f}")


Accuracy of KNeighborsClassifier with K=3: 0.80


In [24]:
# Task 2.4 Use StandardScaler to scale the data and fit a new KNeighborsClassifier model
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model2 = KNeighborsClassifier(n_neighbors=3)
model2.fit(X_train_scaled, y_train)

y_pred2 = model2.predict(X_test_scaled)
accuracy2 = accuracy_score(y_test, y_pred2)
print(f"Accuracy of KNeighborsClassifier with K=3 and StandardScaler: {accuracy2:.2f}")


Accuracy of KNeighborsClassifier with K=3 and StandardScaler: 0.80


In [25]:
# Task 2.5 Use MinMaxScaler to scale the data and fit a new KNeighborsClassifier model
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model3 = KNeighborsClassifier(n_neighbors=3)
model3.fit(X_train_scaled, y_train)

y_pred3 = model3.predict(X_test_scaled)
accuracy3 = accuracy_score(y_test, y_pred3)
print(f"Accuracy of KNeighborsClassifier with K=3 and MinMaxScaler: {accuracy3:.2f}")

Accuracy of KNeighborsClassifier with K=3 and MinMaxScaler: 0.80


In [26]:
# Task 2.6 Evaluate the models with different values of K
for k in range(1, 6):
    model = KNeighborsClassifier(n_neighbors=k)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Accuracy of KNeighborsClassifier with K={k}: {accuracy:.2f}")

    model2 = KNeighborsClassifier(n_neighbors=k)
    model2.fit(X_train_scaled, y_train)
    y_pred2 = model2.predict(X_test_scaled)
    accuracy2 = accuracy_score(y_test, y_pred2)
    print(f"Accuracy of KNeighborsClassifier with K={k} and StandardScaler: {accuracy2:.2f}")

    model3 = KNeighborsClassifier(n_neighbors=k)
    model3.fit(X_train_scaled, y_train)
    y_pred3 = model3.predict(X_test_scaled)
    accuracy3 = accuracy_score(y_test, y_pred3)
    print(f"Accuracy of KNeighborsClassifier with K={k} and MinMaxScaler: {accuracy3:.2f}")

Accuracy of KNeighborsClassifier with K=1: 0.90
Accuracy of KNeighborsClassifier with K=1 and StandardScaler: 0.90
Accuracy of KNeighborsClassifier with K=1 and MinMaxScaler: 0.90
Accuracy of KNeighborsClassifier with K=2: 0.70
Accuracy of KNeighborsClassifier with K=2 and StandardScaler: 0.70
Accuracy of KNeighborsClassifier with K=2 and MinMaxScaler: 0.70
Accuracy of KNeighborsClassifier with K=3: 0.80
Accuracy of KNeighborsClassifier with K=3 and StandardScaler: 0.80
Accuracy of KNeighborsClassifier with K=3 and MinMaxScaler: 0.80
Accuracy of KNeighborsClassifier with K=4: 0.70
Accuracy of KNeighborsClassifier with K=4 and StandardScaler: 0.70
Accuracy of KNeighborsClassifier with K=4 and MinMaxScaler: 0.70
Accuracy of KNeighborsClassifier with K=5: 0.70
Accuracy of KNeighborsClassifier with K=5 and StandardScaler: 0.70
Accuracy of KNeighborsClassifier with K=5 and MinMaxScaler: 0.70


#Task 2.7 

The model that obtains the highest level of accuracy on the test data is considered to be the most precise, and in this particular case, that would be the model with k = 1.

Student name: Khalid Nimri

Student ID: 2140145